src/internal/connector/exchange/data_collections.go

```diff
@@ -16,51 +17,78 @@ import (
 func MetadataFileNames(cat path.CategoryType) []string {
 	switch cat {
 	case path.EmailCategory, path.ContactsCategory:
-		return []string{graph.DeltaTokenFileName, graph.PreviousPathFileName}
+		return []string{graph.DeltaURLsFileName, graph.PreviousPathFileName}
 	default:
 		return []string{graph.PreviousPathFileName}
 	}
 }
 
-// ParseMetadataCollections produces two maps:
-// 1- paths: folderID->filePath, used to look up previous folder pathing
-// in case of a name change or relocation.
-// 2- deltas: folderID->deltaToken, used to look up previous delta token
-// retrievals.
+type CatDeltaPaths map[path.CategoryType]DeltaPaths
+
+type DeltaPaths struct {
+	deltas map[string]string
+	paths  map[string]string
+}
+
+func makeDeltaPaths() DeltaPaths {
+	return DeltaPaths{
+		deltas: map[string]string{},
+		paths:  map[string]string{},
+	}
+}
+
+// ParseMetadataCollections produces a map of structs holding delta
+// and path lookup maps.
 func ParseMetadataCollections(
 	ctx context.Context,
 	colls []data.Collection,
-) (map[string]string, map[string]string, error) {
-	var (
-		paths  = map[string]string{}
-		deltas = map[string]string{}
-	)
+) (CatDeltaPaths, error) {
+	cdp := CatDeltaPaths{
+		path.ContactsCategory: makeDeltaPaths(),
+		path.EmailCategory:    makeDeltaPaths(),
+		path.EventsCategory:   makeDeltaPaths(),
+	}
 
 	for _, coll := range colls {
-		items := coll.Items()
+		var (
+			breakLoop bool
+			items     = coll.Items()
+			category  = coll.FullPath().Category()
+		)
 
 		for {
-			var breakLoop bool
-
 			select {
 			case <-ctx.Done():
-				return nil, nil, errors.Wrap(ctx.Err(), "parsing collection metadata")
+				return nil, errors.Wrap(ctx.Err(), "parsing collection metadata")
+
 			case item, ok := <-items:
 				if !ok {
 					breakLoop = true
 					break
 				}
 
+				m := map[string]string{}
+				cdps := cdp[category]
+
 				switch item.UUID() {
-				// case graph.PreviousPathFileName:
-				case graph.DeltaTokenFileName:
-					err := json.NewDecoder(item.ToReader()).Decode(&deltas)
+				case graph.DeltaURLsFileName:
+					err := json.NewDecoder(item.ToReader()).Decode(&m)
 					if err != nil {
-						return nil, nil, errors.New("parsing delta token map")
+						return nil, errors.New("parsing delta token map")
 					}
 
-					breakLoop = true
+					cdps.deltas = common.UnionMaps(cdps.deltas, m)
+
+				case graph.PreviousPathFileName:
+					err := json.NewDecoder(item.ToReader()).Decode(&m)
+					if err != nil {
+						return nil, errors.New("parsing previous path map")
+					}
+
+					cdps.paths = common.UnionMaps(cdps.paths, m)
```

Comment:
this kind of worries me because it seems like it's going to make it harder to find bugs

Under the current implementation of selectors it's technically possible for GC to be backing up multiple resource owners at the same time. This is saying that we are going to make a single large map containing all the delta URLs and path mappings for all resource owners

Under the expected updated version of selectors, we would have a 1:1 mapping between resource owner and other filters/includes, but then it should be an error to have multiple different "base" snapshots for a single (resource owner, service, category) tuple. It's possible to have multiple bases for a (resource owner, service) pair, but once the category is added it should be 1:1, otherwise we are essentially pulling data from multiple different points in time and that will lead to correctness bugs

Reply:
> Under the current implementation of selectors it's technically possible for GC to be backing up multiple resource owners at the same time.

A bit pedantic, but: although we're looking to selectors for part of this fix, this is less of a selectors issue to enforce than it is a GC/operations layer enforcement.

> This is saying that we are going to make a single large map containing all the delta URLs and path mappings for all resource owners

We're already guaranteeing that isolation from the CLI, and although the sdk api is awaiting that enforcement as well our only known sdk consumer is known to work on a per-resource-owner basis.  I don't think there's much to stress about here.

In either case, the larger concern might be memory issues.  A correctness bug would require duplicate folderIDs across resource owners.  Unlikely, at best.

Reply:
>A correctness bug would require duplicate folderIDs across resource owners. Unlikely, at best.

Actually there's another correctness issue: we could end up accidentally pulling in metadata from multiple base snapshots and that metadata could overlap in some places. I don't think there's currently code to handle this anywhere in the codebase (I've been working on a patch for KW to handle this when building the hierarchy). Under that condition, it would be possible to end up with two sets of delta tokens and path mappings for a category, one from each of the bases. It would be incorrect to use both and it's also incorrect to have some merge algorithm for them. We should only ever use all of one or all of the other for any given (resource owner, service, category) tuple

To make this more concrete, consider my new favorite example:
```text
// Full backup, no base or metadata
backup create exchange user1 email,contacts -> B1

// Should use B1 to source email metadata
backup create exchange user1 email -> B2

// Should use B1 to source contacts metadata and B2 to source email metadata
backup create exchange user1 email,contacts
```

If we just pull all the metadata for all the base backups and merge the results then we'll end up with some mixture of B1 and B2 for the starting point of email when making the final backup. That will be a correctness issue and I think it would be undefined behavior with the current code since the outcome is dependent on iteration order in some places as well as how the M365 hierarchy has changed

Reply:
Ah, gotcha.  That makes sense.  So, for the purposes of metadata, is there ever a situation where we want the consumer to contend with this state?  Forwardly: is this a consumer concern, or should it be a producer concern?  If it's necessary to only consume a certain subset of the metadata, perhaps kopia should only return that set.

Reply:
> If it's necessary to only consume a certain subset of the metadata, perhaps kopia should only return that set.

Since we're using the `Restore` API in KW, implementing this would require pushing extra logic into KW and I'd like to avoid that.

It would be possible for BackupOp to weed out some of this, and maybe that's the best we can do. The code would be rather generic and would make use of the `Reason` field that was recently added to returned snapshots (#1779)